services/web-app/pages/_middleware.js

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright IBM Corp. 2022, 2022
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { NextResponse } from 'next/server'
+
+import { isValidIbmEmail } from '@/utils/string'
+
+import { getRunMode, RunMode } from '../../../packages/api/runtime'
+
+/**
+ * Creates request options for retrieveUser api call including headers and agent if necessary
+ *
+ * @param {NextRequest} req getServerSideProps request object
+ * @returns {RequestInit} request options
+ */
+function getRequestOptions(req) {
+  let cookies = ''
+  Object.entries(req.cookies).forEach(([key, val]) => {
+    cookies += `${key}=${val}; `
+  })
+  return {
+    headers: { cookie: cookies }
+  }
+}
+export async function middleware(req) {
+  if (req.nextUrl.pathname === '/404' || req.nextUrl.pathname.startsWith('/api/')) {
+    return NextResponse.next()
+  }
+  const protocol =
+    getRunMode === RunMode.Prod || process.env.RUNNING_SECURELY === '1' ? 'https' : 'http'
+  const userResponse = await fetch(
+    `${protocol}://localhost:${req.nextUrl.port}/api/user`,
+    getRequestOptions(req)
+  )
+  let user
+  if (userResponse.ok) {
+    user = await userResponse.json()
+  }
+  if (!isValidIbmEmail(user?.email ?? '')) {
+    const url = req.nextUrl.clone()
+    url.pathname = '/404'
+    return NextResponse.rewrite(url)
+  }
+  return NextResponse.next()
```

Comment:
```suggestion
  // Guard - non-200 user api response
  if (!userResponse.ok) {
    return exitWith404(req)
  }
  
  const user = await userResponse.json()
  
  // Guard - no valid user
  if (!user?.email) {
    return exitWith404(req)
  }
  
  // Guard - not a valid IBMer
  if (!isValidIbmEmail(user.email)) {
    return exitWith404(req)
  }
  
  return NextResponse.next()
```

My general philosophy with permissions model and security stuff is to keep the logic as flat as possible with the final "success" statement at the very bottom of the function. (Ideally no more than +1 cognitive complexity for the entire method). Each check or "guard" above that final line should handle one possible error scenario that prevents the success statement from being reached.